### PR TITLE
Axios Hooks

### PR DESCRIPTION
**Pull request by taitfuller**
Implements `useGet()` and `usePost()` hooks that use Axios internally.

### Background

These hooks manage request state for loading and error handling. The hooks expose this state in addition to a function to invoke a request by returning the following type:

```typescript
[
  state: {
    isLoading: boolean,
    error: AxiosError | Error | undefined
  },
  request: (controller?: AbortController) => Promise<AxiosRequest<T, D> | null>
] = useGet<T, D>() / usePost<T, D>()
```

The `request` function returns a `Promise` that resolves to either an `AxiosRequest` or `null` if an error occurs. It also has an optional parameter for `controller`. This is explained in more detail in the [Axios docs](https://axios-http.com/docs/cancellation), but the TLDR is that the `controller` provides the means to cancel a request before it has resolved. This is important when requests are called inside a `useEffect()` to perform [proper cleanup](https://reactjs.org/docs/hooks-effect.html#example-using-hooks-1).

The generic `T` specifies the return type of the response body. Conversely `D` specifies the type of the request body.

#### `useGet()`

The `useGet()` hook has the following call signature consisting of the `url` to use and an optional [Axios Request `config` object](https://axios-http.com/docs/req_config):

```typescript
useGet<T, D>(url: string, config?: AxiosRequestConfig<D>)
``` 

#### `usePost()`

The `usePost()` hook has a similar call signature to `useGet()` however includes an optional `data` parameter which is used for the `body` of the HTTP request:

```typescript
usePost<T, D>(url: string, data?: D, config?: AxiosRequestConfig<D>)
```

### Note

This PR also deletes `src/api/`, which is made redundant by these changes.

### CODE REVIEW

File: frontend/src/pages/NicknamePage/index.tsx [hunk 1]
```diff
@@ -17,25 +22,49 @@ type Props = {
...
   const [token, setToken] = useTokenState("");
   const browserHistory = useContext(BrowserHistoryContext);
 
-  const tryConnect = useCallback(async () => {
-    const res = await validateGame({ gameCode });
+  const [, validateGameCode] = useGet(
+    "/api/game/validate",
+    /*
+    As `validateGameCode` runs inside a `useEffect()` we need to create the
+    config object with `useMemo()` to prevent necessary re-renders.
```
Comment (Alxjlv) [inline frontend/src/pages/NicknamePage/index.tsx:33]:
```suggestion
    config object with `useMemo()` to prevent unnecessary re-renders.
```

**Review by Alxjlv — APPROVED** [review]
Looks good but it has also veered into the "I don't understand how most of this actually works" territory. I'd recommend putting some of your explanation in the PR body into comments in the axios.ts class

**Comment by taitfuller** [comment]
I've updated the documentation in ae5fe20 - let me know what you think!

**Review by Alxjlv — APPROVED** [review]
New documentation looks solid

**Pull request merged**